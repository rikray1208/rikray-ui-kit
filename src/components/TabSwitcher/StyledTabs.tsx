import styled, { css } from 'styled-components';

import { TabsProps } from './index';

import { setFontSize } from '../../utils/helpers';
import { Size } from '../types';

const listReset = css`
    margin: 0;
    padding: 0;
    text-indent: 0;
    list-style-type: none;
`;

function setIconSize(size: Size) {
    switch (size) {
        case 'Large':
            return css`
                height: ${({ theme }) => theme.TabSwitcher['fontSize-xl'] + 'rem'};
            `;
        case 'Medium':
            return css`
                height: ${({ theme }) => theme.TabSwitcher['fontSize-lg'] + 'rem'};
            `;
        case 'Small':
            return css`
                height: ${({ theme }) => theme.TabSwitcher['fontSize-base'] + 'rem'};
            `;
    }
}

export const TabsWrapper = styled.div`
    ${({ theme }) => css`
        width: 100%;
        height: 100%;

        display: flex;
        flex-direction: column;
        gap: ${theme.TabSwitcher['padding-4'] + 'rem'};
    `}
`;

export const TabsList = styled.ul<Pick<TabsProps, 'centered'>>`
    ${listReset}

    ${({ theme, centered }) => css`
        display: flex;
        justify-content: ${centered ? 'space-around' : 'flex-start'};
        gap: ${theme.TabSwitcher['padding-4'] + 'rem'};
        border-bottom: ${theme.TabSwitcher.borderWidth} ${theme.TabSwitcher.borderStyle}
            ${theme.TabSwitcher.colorBorderSecondary};
    `}
`;

export const TabItem = styled.li<Pick<TabsProps, 'size' | 'bold'>>`
    ${listReset}

    ${({ theme, size, bold }) => css`
        display: flex;
        align-items: center;
        gap: ${theme.TabSwitcher['padding-1'] + 'rem'};
        padding-bottom: ${theme.TabSwitcher['padding-2'] + 'rem'};

        cursor: pointer;
        ${setFontSize(size ? size : 'Medium', 'TabSwitcher', { greater: true })}
        font-weight: ${bold && theme.TabSwitcher.fontWeightBold};

        &:hover {
            color: ${theme.TabSwitcher.colorHoverPrimary};
        }

        &.Active {
            color: ${theme.TabSwitcher.colorPrimary};
            border-bottom: ${theme.TabSwitcher.borderWidth} ${theme.TabSwitcher.borderStyle}
                ${theme.TabSwitcher.colorPrimary};
        }

        > svg {
            ${setIconSize(size ? size : 'Medium')}
        }
    `}
`;
