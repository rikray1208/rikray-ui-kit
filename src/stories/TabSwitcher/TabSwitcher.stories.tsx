import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Album, Check, Info } from 'lucide-react';
import React from 'react';

import { Card, TabSwitcher, ThemeProvider } from '../../index';

const decorators = [
    (Story: any) => (
        <ThemeProvider>
            <Card style={{ width: 400 }}>
                <Story />
            </Card>
        </ThemeProvider>
    ),
];

const Items = [
    {
        id: 0,
        header: 'First',
        content: <p style={{ margin: 0 }}>This is the first item.</p>,
    },
    {
        id: 1,
        header: 'Second',
        content: <p style={{ margin: 0 }}>This is the second item.</p>,
    },
    {
        id: 2,
        header: 'Third',
        content: <p style={{ margin: 0 }}>This is the third item.</p>,
    },
];
const ItemsWithIcons = [
    {
        id: 0,
        header: 'First',
        icon: <Album />,
        content: <p style={{ margin: 0 }}>This is the first item.</p>,
    },
    {
        id: 1,
        header: 'Second',
        icon: <Info />,
        content: <p style={{ margin: 0 }}>This is the second item.</p>,
    },
    {
        id: 2,
        header: 'Third',
        icon: <Check />,
        content: <p style={{ margin: 0 }}>This is the third item.</p>,
    },
];

export default {
    title: 'Design System/Global/TabSwitcher',
    component: TabSwitcher,
    decorators: decorators,
} as ComponentMeta<typeof TabSwitcher>;

const Template: ComponentStory<typeof TabSwitcher> = (args) => <TabSwitcher {...args} />;

export const BaseTabSwitcher = Template.bind({});
BaseTabSwitcher.args = {
    items: Items,
    defaultActiveTabId: 2,
    onChange: (tab) => console.log('@active tab:', tab),
};

export const TabSwitcherWithIcons = Template.bind({});
TabSwitcherWithIcons.args = {
    items: ItemsWithIcons,
    defaultActiveTabId: 2,
};
