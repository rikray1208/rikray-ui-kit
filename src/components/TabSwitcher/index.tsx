import React, { FC, useState } from 'react';

import { TabItem, TabsList, TabsWrapper } from './StyledTabs';

import type { Size } from '../types';
import type { LucideIcon } from 'lucide-react';

interface TabItem {
    id: number;
    header: string;
    content: React.ReactNode;
    icon?: React.ReactElement<LucideIcon>;
}
export interface TabsProps {
    items: TabItem[];
    onChange?: (tab: TabItem) => void;
    defaultActiveTabId?: number;
    centered?: boolean;
    size?: Size;
    bold?: boolean;
    className?: string;
    style?: React.CSSProperties;
}
const TabSwitcher: FC<TabsProps> = (props) => {
    const { items, defaultActiveTabId, onChange, centered, bold, size, ...defProps } = props;

    const [activeID, setActiveID] = useState(defaultActiveTabId ?? items[0].id);
    const activeTab = items.find((tab) => tab.id === activeID);

    function handleOnClick(tab: TabItem) {
        setActiveID(tab.id);
        onChange && onChange(tab);
    }

    return (
        <TabsWrapper {...defProps}>
            <TabsList centered={centered}>
                {items.map((tab) => (
                    <TabItem
                        key={tab.id}
                        onClick={() => handleOnClick(tab)}
                        className={activeID === tab.id ? 'Active' : ''}
                        bold={bold}
                        size={size}
                    >
                        {tab.icon && tab.icon}
                        {tab.header}
                    </TabItem>
                ))}
            </TabsList>
            {activeTab?.content}
        </TabsWrapper>
    );
};

export default TabSwitcher;
