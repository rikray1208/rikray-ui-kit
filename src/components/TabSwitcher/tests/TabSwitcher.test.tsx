import React from 'react';
import { create } from 'react-test-renderer';

import { ThemeProvider } from '../../../index';

import TabSwitcher from '../index';

const MockItems = [
    {
        id: 0,
        header: 'First',
        content: <p style={{ margin: 0 }}>This is the first item.</p>,
    },
    {
        id: 1,
        header: 'Second',
        content: <p style={{ margin: 0 }}>This is the second item.</p>,
    },
    {
        id: 2,
        header: 'Third',
        content: <p style={{ margin: 0 }}>This is the third item.</p>,
    },
];

describe('TabSwitcher', () => {
    it('should return a TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a centered TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} centered />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a bold TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} bold />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a large TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} size={'Large'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a Medium TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} size={'Medium'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a Small TabSwitcher', () => {
        const element = (
            <ThemeProvider>
                <TabSwitcher items={MockItems} size={'Small'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
});
