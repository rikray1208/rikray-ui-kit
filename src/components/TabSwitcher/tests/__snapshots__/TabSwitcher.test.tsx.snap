// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabSwitcher should return a Medium TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk dDEEna"
  >
    <li
      className="sc-cabOPr iATa-Dz Active"
      onClick={[Function]}
      size="Medium"
    >
      First
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
      size="Medium"
    >
      Second
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
      size="Medium"
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;

exports[`TabSwitcher should return a Small TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk dDEEna"
  >
    <li
      className="sc-cabOPr fpvFsY Active"
      onClick={[Function]}
      size="Small"
    >
      First
    </li>
    <li
      className="sc-cabOPr fpvFsY"
      onClick={[Function]}
      size="Small"
    >
      Second
    </li>
    <li
      className="sc-cabOPr fpvFsY"
      onClick={[Function]}
      size="Small"
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;

exports[`TabSwitcher should return a TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk dDEEna"
  >
    <li
      className="sc-cabOPr iATa-Dz Active"
      onClick={[Function]}
    >
      First
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
    >
      Second
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;

exports[`TabSwitcher should return a bold TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk dDEEna"
  >
    <li
      className="sc-cabOPr fpCetm Active"
      onClick={[Function]}
    >
      First
    </li>
    <li
      className="sc-cabOPr fpCetm"
      onClick={[Function]}
    >
      Second
    </li>
    <li
      className="sc-cabOPr fpCetm"
      onClick={[Function]}
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;

exports[`TabSwitcher should return a centered TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk nFVRi"
  >
    <li
      className="sc-cabOPr iATa-Dz Active"
      onClick={[Function]}
    >
      First
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
    >
      Second
    </li>
    <li
      className="sc-cabOPr iATa-Dz"
      onClick={[Function]}
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;

exports[`TabSwitcher should return a large TabSwitcher 1`] = `
<div
  className="sc-cjibBx jnFkqj"
>
  <ul
    className="sc-jcMfQk dDEEna"
  >
    <li
      className="sc-cabOPr mfxPU Active"
      onClick={[Function]}
      size="Large"
    >
      First
    </li>
    <li
      className="sc-cabOPr mfxPU"
      onClick={[Function]}
      size="Large"
    >
      Second
    </li>
    <li
      className="sc-cabOPr mfxPU"
      onClick={[Function]}
      size="Large"
    >
      Third
    </li>
  </ul>
  <p
    style={
      {
        "margin": 0,
      }
    }
  >
    This is the first item.
  </p>
</div>
`;
